File: bench/nqueens/nqueens.js
// N-Queens: recursive backtracking with an O(row) safety check.
const n = 11;
const cols = new Array(n).fill(0);

function safe(cols, row, col) {
  for (let i = 0; i < row; i++) {
    const c = cols[i];
    if (c === col) {
      return false;
    }
    if (c - i === col - row) {
      return false;
    }
    if (c + i === col + row) {
      return false;
    }
  }
  return true;
}

function solve(cols, row, n) {
  if (row === n) {
    return 1;
  }
  let count = 0;
  for (let col = 0; col < n; col++) {
    if (safe(cols, row, col)) {
      cols[row] = col;
      count += solve(cols, row + 1, n);
    }
  }
  return count;
}

console.log(solve(cols, 0, n));

File: bench/nqueens/nqueens.lua
-- N-Queens: recursive backtracking with an O(row) safety check.
local n = 11
local cols = {}
for i = 0, n - 1 do
    cols[i] = 0
end

local function safe(cols, row, col)
    for i = 0, row - 1 do
        local c = cols[i]
        if c == col then
            return false
        end
        if c - i == col - row then
            return false
        end
        if c + i == col + row then
            return false
        end
    end
    return true
end

local function solve(cols, row, n)
    if row == n then
        return 1
    end
    local count = 0
    for col = 0, n - 1 do
        if safe(cols, row, col) then
            cols[row] = col
            count = count + solve(cols, row + 1, n)
        end
    end
    return count
end

print(solve(cols, 0, n))
